File: src/main/java/com/nospawnn/ui/argeditors/ArgumentEditor.java
package com.nospawnn.ui.argeditors;

import com.nospawnn.ui.listeners.ArgListener;

import javax.swing.*;
import java.awt.*;

abstract public class ArgumentEditor extends JPanel {
    abstract public Object getValue();

    abstract public JLabel getLabel();

    abstract public String getLabelText();

    abstract public Component getEditor();

    abstract public JPanel getContentPane();

    abstract public void addListener(ArgListener al);

    abstract public void notifyListeners();
}

File: src/main/kotlin/ui/argumenteditors/ArgumentEditor.kt
package com.nospawnn.ui.argumenteditors

import java.awt.Component
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import javax.swing.JLabel
import javax.swing.JPanel

abstract class ArgumentEditor(argName: String) : JPanel() {
    private val label = JLabel(argName)
    abstract val default: Any
    abstract val editor: Component

    private val gbc = GridBagConstraints().apply {
        fill = GridBagConstraints.CENTER
        gridx = GridBagConstraints.RELATIVE
        anchor = GridBagConstraints.WEST
    }

    init {
        layout = GridBagLayout()
        gbc.gridy = 0
        add(label, gbc)
    }

    fun addEditor(editor: Component, newLine: Boolean = false) {
        if (newLine) gbc.gridy = 1
        add(editor, gbc)
    }

    abstract fun getValue(): Any
}